orderer/consensus/follower/follower_chain.go

```diff
@@ -257,14 +331,92 @@ func (c *Chain) pull() error {
 // pullUpToJoin pulls blocks up to the join-block height without inspecting membership on fetched config blocks.
 // It checks whether the chain was stopped between blocks.
 func (c *Chain) pullUpToJoin() error {
-	//TODO
-	return errors.New("not implemented yet: pull up to join block")
+	targetHeight := c.joinBlock.Header.Number + 1
+	firstBlockToPull := c.support.Height()
+	c.logger.Debugf("first block to pull: %d, target height: %d", firstBlockToPull, targetHeight)
+	if firstBlockToPull >= targetHeight {
+		c.logger.Infof("Target height (%d) is <= to our ledger height (%d), skipping pulling", targetHeight, firstBlockToPull)
+		return nil
+	}
+
+	puller, err := c.createBlockPullerFunc()
+	if err != nil {
+		c.logger.Errorf("Error creating block puller: %s", err)
+		return errors.Wrap(err, "error creating block puller")
+	}
+	defer puller.Close()
+
+	var nextBlock *common.Block
+	var actualPrevHash []byte
+
+	// Pull the first block
+	select {
+	case <-c.stopChan:
+		c.logger.Warnf("Stopped before pulling blocks: first to pull was: %d, target height was: %d", firstBlockToPull, targetHeight)
+		return errors.New("stopped before pulling blocks")
+	default:
+		if firstBlockToPull == 0 {
+			nextBlock = puller.PullBlock(firstBlockToPull)
+		} else {
+			prevBlock := c.support.Block(firstBlockToPull - 1)
+			if prevBlock == nil {
+				return errors.Errorf("cannot retrieve previous block %d", firstBlockToPull-1)
+			}
+			actualPrevHash = protoutil.BlockHeaderHash(prevBlock.Header)
+			nextBlock = puller.PullBlock(firstBlockToPull)
+		}
+	}
+
+	if nextBlock == nil {
+		return errors.Wrapf(cluster.ErrRetryCountExhausted, "failed to pull first block %d", firstBlockToPull)
+	}
+	c.logger.Debugf("Pulled first block %d", nextBlock.Header.Number)
+
+	reportedPrevHash := nextBlock.Header.PreviousHash
+	if len(actualPrevHash) > 0 && !bytes.Equal(reportedPrevHash, actualPrevHash) {
+		return errors.Errorf("block header mismatch on sequence %d, expected %x, got %x",
+			nextBlock.Header.Number, actualPrevHash, reportedPrevHash)
+	}
+
+	actualPrevHash = protoutil.BlockHeaderHash(nextBlock.Header)
+	if err = c.support.Append(nextBlock); err != nil {
+		return errors.Wrapf(err, "failed to append block %d to the ledger", nextBlock.Header.Number)
+	}
+	c.resetRetryInterval()
+
+	// Pull the rest of the blocks
+	for seq := uint64(firstBlockToPull + 1); seq < targetHeight; seq++ {
+		select {
+		case <-c.stopChan:
+			c.logger.Warnf("Stopped before pulling all the blocks: pulled blocks from %d until %d, target was %d",
+				firstBlockToPull, nextBlock.Header.Number, targetHeight)
+			return errors.New("stopped before pulling all the blocks")
+		default:
+			nextBlock = puller.PullBlock(seq)
+			if nextBlock == nil {
+				return errors.Wrapf(cluster.ErrRetryCountExhausted, "failed to pull block %d", seq)
+			}
+			reportedPrevHash := nextBlock.Header.PreviousHash
+			if !bytes.Equal(reportedPrevHash, actualPrevHash) {
```

Comment:
I thought the block puller verifies the hash chain on its own? Why do you also do it here?

Reply:
The block puller verifies only batches, there is no continuity between them. I following the pattern from here, which does the same:

https://github.com/hyperledger/fabric/blob/5362a5701d0118629d6b974327555e4769e4232b/orderer/common/cluster/replication.go#L212

Reply:
But we verify the signature, don't we? So how can the signature be good, but the hash not be good?

Reply:
Actually... this internal function is used by 
https://github.com/hyperledger/fabric/blob/84396aaa4164a0d992b3b3e61e55b6ad7a9878c8/orderer/common/cluster/replication.go#L182

which pulls application channels as well as the system channel. I addition, verification is orthogonal (pluggable) and the type of verifier normally used
https://github.com/hyperledger/fabric/blob/84396aaa4164a0d992b3b3e61e55b6ad7a9878c8/orderer/common/cluster/util.go#L202

does not verify all the signatures.

I think the long term solution is to define a "cluster.StrictVerifyBlocks()" that does everything we need. However, I would like to do this under a separate story, after the dust settles on the main join story.

Reply:
Opened https://jira.hyperledger.org/browse/FAB-18104 for this issue